accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1212 @@
+that have `S` as a superinterface, the invocation of `m` is treated as an
+invocation of a regular class instance member whose member signature is the
+combined member signatures of all declarations of `m` in the direct
+superinterfaces of `V`. 
+
+*In other words, members "inherited" from non-extension type
+superinterfaces are invoked as normal class instance members, as if we
+could "see through the veil" that is the extension type and call members of
+the representation type which have been unveiled by including `S` as a
+superinterface, directly or indirectly.*
+
+Otherwise, let _Dm_ be the declaration of `m` that `V` has.
+
+If _Dm_ is a getter declaration with return type `R` then the static
+type of the invocation is
+<code>[T<sub>1</sub>/X<sub>1</sub> .. T<sub>s</sub>/X<sub>s</sub>]R</code>.
+
+If _Dm_ is a method with function type `F`, and `args` is omitted, the
+invocation has static type
+<code>[T<sub>1</sub>/X<sub>1</sub> .. T<sub>s</sub>/X<sub>s</sub>]F</code>.
+*This is an extension type method tear-off.*
+
+If _Dm_ is a method with function type `F`, and `args` exists, the static
+analysis of the extension type member invocation is the same as that of an
+invocation with argument part `args` of a function with type
+<code>[T<sub>1</sub>/X<sub>1</sub> .. T<sub>s</sub>/X<sub>s</sub>]F</code>.
+*This determines the compile-time errors, if any, and it determines the
+type of the invocation as a whole.*
+
+
+### Dynamic Semantics of an Extension Type Member Invocation
+
+Consider an invocation of the extension type member `m` on the receiver `e`
+according to the extension type declaration `V` and with actual type
+arguments <code>T<sub>1</sub>, ..., T<sub>s</sub></code>. If the invocation
+includes an actual argument part (possibly including some actual type
+arguments) then call it `args`. Assume that `V` declares the type variables
+<code>X<sub>1</sub>, ..., X<sub>s</sub></code>.
+
+Let _Dm_ be the declaration named `m` that `V` has.
+
+Evaluation of this invocation proceeds by evaluating `e` to an object
+`o`.
+
+Then, if `args` is omitted and _Dm_ is a getter, execute the body of
+said getter in an environment where `this` and the name of the
+representation are bound to `o`, and the type variables of `V` are
+bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>.
+If the body completes returning an object `o2` then the invocation
+evaluates to `o2`. If the body throws an object and a stack trace
+then the invocation completes throwing the same object and stack
+trace.
+
+Otherwise, if `args` is omitted and _Dm_ is a method, the invocation
+evaluates to a closurization of _Dm_ where
+`this` and the name of the representation are bound to `o`, and the
+type variables of `V` are bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>.
+The operator `==` of the closurization returns true if and only if the
+operand is the same object.
```

Comment:
It follows: If two function objects are the same object then they are equal (we certainly can't have a usable `==` if this is not true). We do not require that the function objects must compare equal if they are, say, tear-offs of the same extension type member with the same representation object.

We cannot require this equality to hold in all cases, because it would be wrong (if they have different type arguments then they certainly shouldn't be `==`), but, just like extension methods, we don't require the equality to hold even in the case where the type arguments are the same.

So the main idea here is: Just like extension methods.

I do not think this prevents optimizations where certain tear-offs are the same ("constant") object even though they are torn off at two different occasions. For example, an extension type with no type arguments and an extension type member that doesn't use `this` could presumably be compiled to a top-level function and reused for all tear-offs of that member.